crates/uv/src/commands/project/mod.rs

```diff
@@ -296,43 +340,64 @@ pub(crate) enum PythonRequestSource {
     /// The request was provided by the user.
     UserRequest,
     /// The request was inferred from a `.python-version` or `.python-versions` file.
-    DotPythonVersion(String),
+    DotPythonVersion(PythonVersionFile),
     /// The request was inferred from a `pyproject.toml` file.
     RequiresPython,
 }
 
+impl std::fmt::Display for PythonRequestSource {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PythonRequestSource::UserRequest => write!(f, "explicit request"),
+            PythonRequestSource::DotPythonVersion(file) => {
+                write!(f, "version file at `{}`", file.path().user_display())
+            }
+            PythonRequestSource::RequiresPython => write!(f, "`requires-python` metadata"),
+        }
+    }
+}
+
 /// The resolved Python request and requirement for a [`Workspace`].
 #[derive(Debug, Clone)]
 pub(crate) struct WorkspacePython {
     /// The source of the Python request.
-    source: PythonRequestSource,
+    pub(crate) source: PythonRequestSource,
     /// The resolved Python request, computed by considering (1) any explicit request from the user
     /// via `--python`, (2) any implicit request from the user via `.python-version`, and (3) any
     /// `Requires-Python` specifier in the `pyproject.toml`.
-    python_request: Option<PythonRequest>,
+    pub(crate) python_request: Option<PythonRequest>,
     /// The resolved Python requirement for the project, computed by taking the intersection of all
     /// `Requires-Python` specifiers in the workspace.
-    requires_python: Option<RequiresPython>,
+    pub(crate) requires_python: Option<RequiresPython>,
 }
 
 impl WorkspacePython {
     /// Determine the [`WorkspacePython`] for the current [`Workspace`].
     pub(crate) async fn from_request(
         python_request: Option<PythonRequest>,
-        workspace: &Workspace,
+        workspace: Option<&Workspace>,
+        project_dir: &Path,
     ) -> Result<Self, ProjectError> {
```

Comment:
Instead of making `workspace` optional here, I might introduce a new abstraction which is used by `WorkspacePython`? It's doing a bit more work than it ought to as-is, but it's not clear another abstraction is worth it.